pex/jobs.py

```diff
@@ -0,0 +1,207 @@
+    **subprocess_kwargs
+  )
+  return Job(command=cmd, process=process)
+
+
+class SpawnedJob(namedtuple('SpawnedJob', ['job', 'result_func'])):
+  """A handle to a spawned :class:`Job` and its associated result."""
+
+  @classmethod
+  def wait(cls, job, result):
+    def wait_result_func():
+      job.wait()
+      return result
+
+    return cls(job=job, result_func=wait_result_func)
+
+  @classmethod
+  def stdout(cls, job, result_func, input=None):
+    def stdout_result_func():
+      stdout, _ = job.communicate(input=input)
+      return result_func(stdout)
+
+    return cls(job=job, result_func=stdout_result_func)
+
+  def await_result(self):
+    """Waits for the spawned job to complete and returns its result."""
+    return self.result_func()
+
+  def kill(self):
+    """Terminates the spawned job if it's not already complete."""
+    self.job.kill()
+
+
+_CPU_COUNT = cpu_count()
+_ABSOLUTE_MAX_JOBS = _CPU_COUNT * 2
+
+
+DEFAULT_MAX_JOBS = _CPU_COUNT
+"""The default maximum number of parallel jobs PEX should use."""
+
+
+def _sanitize_max_jobs(max_jobs=None):
+  assert max_jobs is None or isinstance(max_jobs, int)
+  if max_jobs is None or max_jobs <= 0:
+    return DEFAULT_MAX_JOBS
+  else:
+    return min(max_jobs, _ABSOLUTE_MAX_JOBS)
+
+
+def execute_parallel(max_jobs, inputs, spawn_func, raise_type):
+  """Execute jobs for the given inputs in parallel.
+
+  :param int max_jobs: The maximum number of parallel jobs to spawn.
+  :param inputs: An iterable of the data to parallelize over `spawn_func`.
+  :param spawn_func: A function taking a single input and returning a :class:`SpawnedJob`.
+  :param raise_type: A type that takes a single string argument and will be used to construct a
+                     raiseable value when any of the spawned jobs errors.
+  :returns: An iterator over the spawned job results as they come in.
+  :raises: A `raise_type` exception if any individual job errors.
+  """
+  size = _sanitize_max_jobs(max_jobs)
+  TRACER.log('Spawning a maximum of {} parallel jobs to process:\n  {}'
+             .format(size, '\n  '.join(map(str, inputs))),
+             V=9)
+
+  stop = Event()
+  job_slots = BoundedSemaphore(value=size)
+  spawned_job_queue = Queue()
+  queue_done_sentinel = object()
+
+  def spawn_jobs():
+    for input in inputs:
+      if not stop.is_set():
+        job_slots.acquire()
+        try:
+          item = spawn_func(input)
+        except Exception as e:
+          item = e
+        finally:
+          spawned_job_queue.put(item)
+    spawned_job_queue.put(queue_done_sentinel)
+
+  spawner = Thread(name='PEX Parallel Job Spawner', target=spawn_jobs)
+  spawner.daemon = True
+  spawner.start()
+
+  error = None
+  while True:
+    item = spawned_job_queue.get()
+    if item is queue_done_sentinel:
+      if error:
+        raise error
+      return
+
+    try:
+      if isinstance(item, Exception):
+        error = item
+      elif error is not None:
+        item.job.kill()
+      else:
```

Comment:
```python
      else:  # i.e. this is not an exception and there were no prior exceptions
```

Reply:
This one I left out. The prior comment seems useful because of spatial closeness, the words and lines are just close enough together to cause confusion:
```python
...
  error = item
elif error is not None:
...
```

The rest of the logic stands on its own after the comment above, i.e.: exception, prior exception, nominal case.

Reply:
Sounds good. With the comment above, I agree with the decision for no comment here.